### PR TITLE
vulnerable-code: Migrate from Jackson to kotlinx.serialization

### PR DESCRIPTION
**Pull request by sschuberth**
For portability and performance evaluation, migrate the independent
VulnerableCode client library to use kotlinx.serialization for
(de-)serialization.

If the migration turns out to be beneficial in terms of performance,
more independent parts of the code base might be migrated.

Signed-off-by: Sebastian Schuberth <sebastian.schuberth@bosch.io>

### CODE REVIEW

**Review by mnonnenmacher — COMMENTED** [review]
Out of curiosity, how are you going to measure the performance impact? Can't you measure it before merging the PR?

File: advisor/src/main/kotlin/advisors/VulnerableCode.kt [hunk 1]
```diff
@@ -91,7 +91,10 @@ class VulnerableCode(
         startTime: Instant
     ): Map<Package, List<AdvisorResult>> {
         val packageMap = packages.associateBy { it.purl }
-        val packageVulnerabilities = service.getPackageVulnerabilities(PackagesWrapper(packageMap.keys))
+
+        // TODO: Remove the "toList()" conversion once https://github.com/Kotlin/kotlinx.serialization/issues/1421 is
+        // resolved.
```
[the remark this answers is not in the document]
Comment (sschuberth) [reply advisor/src/main/kotlin/advisors/VulnerableCode.kt:96]: As the code will fail to compile when removing `toList()` here, I thought it's fine to have the comment just here, as the other place then cannot be overlooked.

[the remark this answers is not in the document]
Comment (sschuberth) [reply advisor/src/main/kotlin/advisors/VulnerableCode.kt]: After having a look again, I was even able to get rid of this code change altogether, and as a result I did indeed move the comment to the `PackagesWrapper` class only.

**Comment by sschuberth** [comment]
> Out of curiosity, how are you going to measure the performance impact? Can't you measure it before merging the PR?

I was planning to monitor it for a variety of projects scanned in production over time.

**Comment by sschuberth** [comment]
FYI: https://www.ericthecoder.com/2020/11/23/benchmarking-kotlin-json-parsers-jackson-kotlin-and-kotlinx-serialization/

**Comment by oheger-bosch** [comment]
> 
> 
> > Out of curiosity, how are you going to measure the performance impact? Can't you measure it before merging the PR?
> 
> I was planning to monitor it for a variety of projects scanned in production over time.

It is probably not straight-forward to gather correct figures. The serialization/deserialization is done behind the scenes by Retrofit. So it is not clear, which time is consumed by the marshalling and by the network communication. As I would expect the latter to be the major portion, measuring becomes even harder.

**Comment by sschuberth** [comment]
> It is probably not straight-forward to gather correct figures. The serialization/deserialization is done behind the scenes by Retrofit. So it is not clear, which time is consumed by the marshalling and by the network communication. As I would expect the latter to be the major portion, measuring becomes even harder.

Your probably right. But I regarding this to be a good change anyway, for the mentioned portability, and to gather experience with kotlinx.serialization. So, any objections to getting this merged?

**Comment by oheger-bosch** [comment]
> You're probably right. But I regard this to be a good change anyway, for the mentioned portability, and to gather experience with kotlinx.serialization. So, any objections to getting this merged?

I will not veto this change. It is, however, confusing to use multiple different serialization mechanisms in the codebase. So, should there be a comment somewhere stating that this is an experiment? Or do you think that the commit message is sufficient?

**Comment by sschuberth** [comment]
> It is, however, confusing to use multiple different serialization mechanisms in the codebase.

Actually, we should not consider anything under the `clients` directory to be part of the ORT codebase in the strict sense, also see the [README](https://github.com/oss-review-toolkit/ort/tree/master/clients). We have a long-pending plan to outsource "clients" (as well as other independent libraries, like "spdx-utils") to separate Git repos (and publish artifacts) for independent use by third-parties.
 
> So, should there be a comment somewhere stating that this is an experiment? Or do you think that the commit message is sufficient?

I believe it to be more than an experiment. In my view using kotlinx.serialization exclusively is the definite way to go, so see [this issue](https://github.com/oss-review-toolkit/ort/issues/3904). In any case, I believe the commit message to be sufficient.

**Review by oheger-bosch — APPROVED** [review]
Thanks for clarifying.

**Pull request merged**